_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+![`Module.new` in `Msf::ModuleManager#reload_module`](/images/module-new-in-msf-module-manager-reload-module.png)
+
+The first one is in `Msf::ModuleManager#reload_module`.  It is unlikely that the only way to load modules is to use a function call `reload`, so moving on...
+
+![`Module.new` in `Msf::ModuleManager#load_module_from_file`](/images/module-new-in-msf-module-manager-load-module-from-file.png)
+
+`Msf::ModuleManager#load_module_from_file` is probably the correct method, but let's check the last line.
+
+![`Module.new` in `Msf::ModuleManager#load_module_from_archive`](/images/module-new-in-msf-module-manager-load-module-from-archive.png)
+
+Huh, so `Msf::ModuleManager#load_module_from_file` and `Msf::ModuleManager#load_module_from_archive` look really similar, except one using `Msf::ModuleManager#load_module_source` and the other is using `Fastlib.load`.  We'll skip over what `Fastlib` is for now and stick with the `Module.new` in `Msf::ModuleManager#load_module_from_path`.  We need to read all of `Msf::ModuleManager#load_module_from_file` to understand the context of the `Module.new` call on line 864.
+
+### De-anonymizing namespace Modules
+
+The minimal change I can make to `Msf::ModuleManager#load_module_from_file` is that I replace the direct call to `Module.new` with a method that returns an named Module instead.  Here's the constraint this new method must fulfill from the code we've seen:
+
+1. The name must be unique, so that it prevent naming collision
+2. It needs to support reloading modules, so the name needs to be deterministic so that it can be replaced
+
+Let's look at [the method I came up with](https://github.com/rapid7/metasploit-framework/blob/8a2dc0a09fbe4ed9880b515d36df751042b09d57/lib/msf/core/module_manager.rb#L1128-L1185)
+
+![`Msf::ModuleManager.wrapper_module`](/images/msf-module-manager-wrapper-module.png)
+
+`name` when it comes in is `/` separated, such as `auxiliary/pro/social_engineering/web_phish`.  I know from Rails that `String#camelize` in `activesupport` can be used to convert a path to a `Module#name`.
+
+![`Msf::ModuleManager.wrapper_module` `name` argument](/images/msf-module-manager-wrapper-module-name.png)
+
+So, `relative_module_name`, is `'Auxiliary::Pro::SocialEngineering::WebPhish'`
+
+![`Msf::ModuleManager.wrapper_module` `relative_module_name` example](/images/msf-module-manager-wrapper-module-relative-module-name.png)
+
+Because this now gives us a hierarchy of namespace modules, we need to ensure that each parent module is created before any child modules
+
+![`Msf::ModuleManager.wrapper_module` `until` loop](/images/msf-module-manager-wrapper-module-until.png)
+
+First, why am I starting with `Object` as the parent?  Well, it's because constant resolution all start with `Object.get_const`, which I got from `ModuleConstMissing#const_missing` (https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L176)
```

Comment:
First, why am I starting with `Object` as the parent?  Well, it's because constant resolution all **~~start~~** **starts** with `Object.get_const`, which I got from `ModuleConstMissing#const_missing` (https://github.com/rails/rails/blob/v3.2.2/activesupport/lib/active_support/dependencies.rb#L176)